src/sentry/integrations/discord/integration.py

```diff
@@ -108,40 +110,85 @@ def __init__(self) -> None:
             "external_id": guild_id,
+            "user_identity": {
+                "type": "discord",
+                "external_id": discord_user_id,
+                "scopes": [],
+                "data": {},
+            },
         }
 
-    def get_guild_name(self, guild_id: str) -> str:
-        url = DiscordClient.GUILD_URL.format(guild_id=guild_id)
+    def setup(self) -> None:
+        if self._credentials_exist():
+            DiscordCommandManager().register_commands()
+
+    def _get_guild_name(self, guild_id: str) -> str:
+        url = self.client.GUILD_URL.format(guild_id=guild_id)
         headers = {"Authorization": f"Bot {self.bot_token}"}
         try:
-            response = DiscordClient().get(url, headers=headers)
-            guild_name = response["name"]  # type:ignore
-        except ApiError:
+            response = self.client.get(url, headers=headers)
+            return response["name"]  # type: ignore
+        except (ApiError, AttributeError):
             return guild_id
-        return guild_name
 
-    def get_bot_install_url(self):
-        setup_url = absolute_uri("extensions/discord/setup/")
+    def _get_discord_user_id(self, auth_code: str) -> str:
+        """
+        Helper function for completing the oauth2 flow and grabbing the
+        installing user's Discord user id so we can link their identities.
 
-        return f"https://discord.com/api/oauth2/authorize?client_id={self.application_id}&permissions={self.bot_permissions}&redirect_uri={setup_url}&response_type=code&scope={' '.join(self.oauth_scopes)}"
+        We don't keep the granted token beyond this function because we don't
+        need it.
 
-    def setup(self) -> None:
-        if self._credentials_exist():
-            DiscordCommandManager().register_commands()
+        If something goes wrong with this we will throw an error because we
+        need an initial identity to configure the identity provider for this
+        integration.
+
+        """
+        form_data = f"client_id={self.application_id}&client_secret={self.client_secret}&grant_type=authorization_code&code={auth_code}&redirect_uri={self.setup_url}"
+        headers = {
+            "Content-Type": "application/x-www-form-urlencoded",
+        }
+
+        try:
+            response = self.client.post(
+                "https://discord.com/api/v10/oauth2/token",
+                json=False,
+                data=form_data,
+                headers=headers,
+            )
+            token = response["access_token"]  # type: ignore
+
+            headers = {"Authorization": f"Bearer {token}"}
+
+            # Can't use self.client.get because that will overwrite our headers
+            # with our bot's authorization
+            response = requests.get(f"{self.client.base_url}/users/@me", headers=headers)
+            return response.json()["id"]
+        except (ApiError, KeyError):
+            logger.error("discord.install.failed_to_get_discord_user_id")
```

Comment:
Do we want to include any additional info about the error/request? you can include a dict `extra` like so:

```
logger.error(
            "discord.install.failed_to_get_discord_user_id",
            extra={
                "key": value
            },
        )
```